utils/zerovec/src/ule/option.rs

```diff
@@ -0,0 +1,193 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::*;
+use core::mem::{self, MaybeUninit};
+
+/// This type is the [`ULE`] type for `Option<T>`.
+///
+/// # Example
+///
+/// ```rust
+/// use zerovec::ZeroVec;
+///
+/// let z = ZeroVec::alloc_from_slice(&[Some('a'), Some('á'), Some('ø'), Some('ł')]);
+///
+/// assert_eq!(z.get(2), Some(Some(('ø'))));
+/// ```
+// Invariants:
+// The MaybeUninit is zeroed when None (bool = false),
+// and is valid when Some (bool = true)
+#[repr(packed)]
+pub struct OptionULE<T>(bool, MaybeUninit<T>);
+
+impl<T: Copy> OptionULE<T> {
+    /// Obtain this as an Option<T>
+    pub fn get(self) -> Option<T> {
+        if self.0 {
+            unsafe {
+                // safety: self.0 is true so the MaybeUninit is valid
+                Some(self.1.assume_init())
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Construct an OptionULE<T> from an equivalent Option<T>
+    pub fn new(opt: Option<T>) -> Self {
+        if let Some(inner) = opt {
+            Self(true, MaybeUninit::new(inner))
+        } else {
+            Self(false, MaybeUninit::zeroed())
+        }
+    }
+}
+
+// Safety (based on the safety checklist on the ULE trait):
+//  1. OptionULE does not include any uninitialized or padding bytes.
+//     (achieved by `#[repr(packed)]` on a struct containing only ULE fields)
```

Comment:
Nit: I think this assertion is not necessarily true since the struct itself does not force the fields to be ULE. Maybe just put the ULE bound onto the struct itself instead of the impl.